### DISCUSSION TITLE
Version extraction for Github Releases

### DISCUSSION THREAD

**Discussion by kjurl** [ImranR98/Obtainium#1286] [Q&A] [answered]
(all the discussion below is in context with Github repo that releases multiple apks in a single release)
I want to know what "Version Extraction RegEx" actually matches to-
1. the link to the apk (right click on assets and copy link - that link)
2. the title of the asset (name of the asset)
3. the release notes
4. the release title 
5. the release tag? (this seems the most accurate assumption)

The repo which I sourced in Obtainium has multiple apks in releases(all apks have different versions), all of their versions mentioned inside the asset-name (and consequently the asset link), and the release notes. 

I can filter out the apk with other options but am stuck when it comes to the version of that apk. How do I get the correct version of the app ?

Some other questions regarding versions that I would also like answered ( if possible :) )0
1. What is the difference between "No version Detection", "Standard Version detection"? when "No version detection" is selected Obtainium still pulls the version (pulling the release tag in my case)?
2. "Standard Version detection" - where does it decide the standard version, what does it do?
3. what happens first filtering of apks(in a multi-apk release), or getting the version?
4. What does the "Use app versionCode as OS-detected version" do? - as I'm not from app-dev background, idk what versionCode and OS-detected version is (some links would be nice)
5. On my device, when I open app-info from an Obtainium installed app, it shows me a version. What does that version depend on?

**Comment by ImranR98** [comment, marked as answer]
Version extraction works differently in the HTML source compared to all others:
- In the HTML source, it is used to extract the version number from the APK URL. If you don't do this, the entire URL is hashed and that hash is used as a "version".
- For all other sources, it is used to remove parts of the version string returned by the source (which, in GitHub's case, is the tag as you mentioned). So for example, if a release's tag is `v1.2.3-fdroid`, you may only want `1.2.3` and can use version detection to accomplish that.

While version extraction is about grabbing a version string from the source (for HTML) or modifying it (for others), version detection is about detecting what version of an app is currently installed on your phone. This is only done for apps where the OS-provided version can be directly compared with the source-provided version string. This is a convenience feature that is optional - turning version detection off will not affect you in most cases. For more details: https://github.com/ImranR98/Obtainium/wiki#version-detection

To answer your third question, the version string for a release provided by a source (like GitHub's tags) have nothing to do with the APKs attached to that release (release assets on GitHub).

"Use app versionCode as OS-detected version" just means that when version detection is enabled, we grab the app's [`versionCode`](https://developer.android.com/studio/publish/versioning#versioningsettings) instead of the normal `versionName`. This is useful if the source provides only versionCodes. Not something you need to worry about for most sources (only seen it in one instance so far).

The "version" you are shown in Obtainium is the version string provided by the source. This does not have to be the actual version of the app when it is installed in the OS - it just needs to be some unique identifier for that release.

**Comment by ImranR98** [comment]
In general, just leave settings at their defaults - they should work most of the time.

**Reply by kjurl** [reply]
Turns out I was over complicating stuff, wanted to see correct version numbers on the Obtanium.

But as you said, "treat version numbers as unique identifiers that are used to compare installed to source app" makes perfect sense. :)
